### ISSUE TITLE
How to add NPM Dependancies for Electron app

### ISSUE THREAD

**Issue by syrys** [NathanWalker/angular-seed-advanced#258]
Hey All,

So, lets say i want to write a electron app, and on `main.desktop.ts` (main electron file), i want to use some node modules. Lets say i want to make a folder copy app using gulp for some reason, so i need the node module "gulp" (use what ever you want, im just using this as an example) and all its dependencies. O

Obviously you install gulp first by using `npm install gulp --save`. Then gulp is in your projects node_modules, great. Then you use `require('gulp')` in `main.desktop.ts` and use gulp magic. All is good, or is it? If you run `npm run start.livesync.desktop.windows` (same with mac), all will be fine. but if you run `npm run build.desktop.windows`, then execute the exe file, you get an error at the above said require statement because gulp was  not found. ok, sure, the electron app doesnt know gulp, fine.

After reading, i found that for some bazaar reason, you have to import node modules one by one defined in `/tools/tasks/project/desktop.libs.ts`. why not the `project.config.ts`? 

Ok, sure, lets add `'node_modules/gulp/**/*',` to `/tools/tasks/project/desktop.libs.ts`, then `npm run build.desktop.windows`. Then run the exe. Oh wait, another dependency missing (`orchestrator`), what is it? Its one of many dependencies of gulp. Grr. Now i have to start adding all the dependencies of all the dependencies into my `desktop.libs.ts`. Wait, then what about all the dependencies of the dependencies of dependencies? What about their dependencies? ... 

You see my problem right? Lot of dependencies have their own dependencies. Adding all of them individually is not really going to work. Is there a better solution? If not, can i propose one?

---

**What i propose** (if no other solution)
- new variable/array available in `project.config.ts`: ELECTRON_NPM_DEPENDANCIES
- you add npm module names into this array, eg: `ELECTRON_NPM_DEPENDANCIES = ['gulp', 'lodash', 'blah'];`
- during the build process `npm run build.desktop.windows`, you add a new task where it creates a `/desktop/windows/resources/app/package.json` (this file is already created as it is, but keep reading). In this file, you add the `ELECTRON_NPM_DEPENDANCIES` array as dependencies. As for the versions of these dependencies (ie. gulp version), you can look this up on the main `/package.json` as gulp should be present on that file as a dependency in the first place.
- once the new package.json file is created and dependencies added, then you run a `npm install` on that newly created package.json to fetch out all the dependencies and their dependencies and so on...

at the end of the day, user simply needs to install the required dependencies using npm, and add the name of the said dependency to `project.config.ts`, thats all. 

Thoughts?

**Comment by syrys** [comment]
So after a bit more experimentation and reading, i realised that the `/desktop/windows/resources/app/package.json` and/or `/src/client/package.json` sort of works the same way as i proposed. That you can add a dependencies section to it and it somehow tries to get them using npm. Example:

```
{
  "name": "TestApp",
  "version": "1.0.0",
  "main": "main.desktop.js",
  "dependencies": {
    "gulp": "^3.9.1"
  }
}
```

What im not 100% sure yet is if this is an automatic function of "electron" or part of the seed project (havent found anything implying its from the seed, so im going to assume its from electron it self). Problem here is, if i try to build with the above package.json, i will still not get all the dependencies required to run gulp, as it doesn't get all gulp's dependencies (i suspect that it sees the current project has them, so it ignores them. an example of a missing dependency is `orchestrator`). That implies that if you have a dependency that is required on the main package.json, then if the same is required in electron's package.json, you are going to have a bad time (specially if that package also has other common dependencies).

Any thoughts on this?

**Comment by kbrandwijk** [comment]
I have made PR #248 recently to add dependencies configured in project.config to desktop.libs automatically, however that just uses whatever is configured in project.config, so if additional dependencies are needed that are not specified there, they will not be included in the desktop build.
The easiest alternative I have found is to do an actual `npm install --save` inside the /src/client folder. This works for dependencies as well, because they are installed into the node_modules folder inside /src/client, and will automatically be copied with the desktop build.

**Comment by geminiyellow** [comment]
@kbrandwijk 

i try to add `angular2-jwt` to this project.

so i add this code in `project.config.ts`

```
    let additionalPackages: ExtendPackages[] = [{
      name: 'angular2-jwt',
      // Path to the package's bundle
      path: 'node_modules/angular2-jwt/angular2-jwt.js',
      packageMeta: { defaultExtension: 'js' }
    }];

    this.addPackagesBundles(additionalPackages);
```
add `"angular2-jwt": "^0.1.28"` to `client/package.json`'s `dependencies`. 
it work in web version , but when i `npm run start.desktop`
got this error:

```
file:///Users/developer/Documents/Sources/client/dist/dev/node_modules/angular2-jwt/angular2-jwt.js Failed to load resource: 
Failed to load resource: net::ERR_FILE_NOT_FOUND
```

so, could you tell me why the dependencies no copy to the `dist` folder ?

**Comment by geminiyellow** [comment]
and, ok, i knew what your PR do is push the `NPM_DEPENDENCIES` to the desktop libs,
and i try it, what the error is said `angular2-jwt.js?1486353097484:32 Uncaught ReferenceError: exports is not defined` so, what i missing? could you help me.

**Comment by Kaffiend** [comment]
just noticed this issue. this was implemented in #352

**Comment by NathanWalker** [comment]
Thanks @Kaffiend !
